### DISCUSSION TITLE
Reduce or manage package dependencies in the API

### DISCUSSION THREAD

**Discussion by heistp** [cue-lang/cue#1388] [Ideas]
**Is your feature request related to a problem? Please describe.**

I'm testing use of the cue API to read in config for an application, as follows (a rough snippet):
```
instances := load.Instances([]string{}, &load.Config{Dir: "."})
for _, i := range instances {
	ctx := cuecontext.New()
	v := ctx.BuildInstance(i)
	if v.Err() != nil {
		log.Fatalf("%s", v.Err())
	}
	fmt.Printf("%s\n", v)
}
```

This does what I expect, but the resulting executable is 15M, with dependencies on many packages I shouldn't actually need, like cockroachdb, protobufs, yaml, etc.

**Describe the solution you'd like**

It would be nice if package dependencies were minimized for applications that are just reading cue files for configuration.

**Describe alternatives you've considered**

One alternative is, I don't use the cue API, and instead ask users to download cue separately and export the config to JSON, which I then import, but this seems less convenient for users than having cue baked in.

**Additional context**

**Comment by verdverm** [comment]
I find this SO answer helpful: https://stackoverflow.com/a/64390523/2366390

List modules in a binary: `go version -m <binary>`

List package in a binary: `go list -deps -json | jq -r 'select(.Module and (.Module.Main | not)) | .Module.Path + "@" + .Module.Version + "\t" + .ImportPath' | sort`

Both will display the list of dependencies for the main package which are actually part of the binary. Generally speaking, Go only includes the packages used into the resulting binary.

I suspect many of the dependencies listed on a simple example like this are due to the load package and handling the various formats. It also looks like the standard library is pulled in so that your users can have access to these features.

Note that _cockroach_ the DB is not included, but rather their Arbitrary Precision Decimal (apd) library is. `cue` will also load Yaml and Json next to your CUE by default, so your users could potentially need this.

For reference, here are the non-cue packages included in [this example](https://cuetorials.com/api/loading-cue/)

```
github.com/cockroachdb/apd/v2@v2.0.1    github.com/cockroachdb/apd/v2
github.com/emicklei/proto@v1.6.15       github.com/emicklei/proto
github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b       github.com/golang/glog
github.com/google/uuid@v1.2.0   github.com/google/uuid
github.com/mpvl/unique@v0.0.0-20150818121801-cbe035fff7de       github.com/mpvl/unique
github.com/pkg/errors@v0.8.1    github.com/pkg/errors
github.com/protocolbuffers/txtpbfmt@v0.0.0-20201118171849-f6a6b3f636fc  github.com/protocolbuffers/txtpbfmt/ast
github.com/protocolbuffers/txtpbfmt@v0.0.0-20201118171849-f6a6b3f636fc  github.com/protocolbuffers/txtpbfmt/parser
github.com/protocolbuffers/txtpbfmt@v0.0.0-20201118171849-f6a6b3f636fc  github.com/protocolbuffers/txtpbfmt/unquote
golang.org/x/net@v0.0.0-20200226121028-0de0cce0169b     golang.org/x/net/idna
golang.org/x/text@v0.3.2        golang.org/x/text/encoding
golang.org/x/text@v0.3.2        golang.org/x/text/encoding/internal
golang.org/x/text@v0.3.2        golang.org/x/text/encoding/internal/identifier
golang.org/x/text@v0.3.2        golang.org/x/text/encoding/unicode
golang.org/x/text@v0.3.2        golang.org/x/text/internal/utf8internal
golang.org/x/text@v0.3.2        golang.org/x/text/runes
golang.org/x/text@v0.3.2        golang.org/x/text/secure/bidirule
golang.org/x/text@v0.3.2        golang.org/x/text/transform
golang.org/x/text@v0.3.2        golang.org/x/text/unicode/bidi
golang.org/x/text@v0.3.2        golang.org/x/text/unicode/norm
golang.org/x/xerrors@v0.0.0-20191204190536-9bdfabe68543 golang.org/x/xerrors
golang.org/x/xerrors@v0.0.0-20191204190536-9bdfabe68543 golang.org/x/xerrors/internal
gopkg.in/yaml.v3@v3.0.0-20200121175148-a6ecf24a6d71     gopkg.in/yaml.v3
```

---

I cannot comment on how feasible it is to reduce this. For comparison on sizes

- `cue` is 19M
- `go` is 14
- `terraform` is 75
- `docker` is 60
- `kubectl` is 42

**Comment by heistp** [comment]
Thanks, yes it's helpful verify the actual dependencies for my executable, which are very similar to the ones you listed from the example.

Even though Go only includes packages used, if a package is used, then it seems that all of it, and its dependencies end up in the executable. I think what may be removed is any code that can't ever be reached.

I guess if YAML or other formats are required for canonical CUE support, those packages would need to be referenced directly, but if not, they could be called through an interface and only registered by executables that need them. I don't envision anyone using anything but JSON and CUE from my app. I'm also not sure how possible it is to reduce the dependencies, but I do think it would be useful to explore...

**Comment by mpvl** [comment]
My suspicion is that the x/text packages will be by far the biggest contributors to the binary size here, but reducing the scope of those will also be quite hard.

Not that `cockroachdb/apd` is just the arbitrary-precision decimal package, which is also crucial to the language.

I think that overall the biggest contributor is the inclusion of the cue stdlib. I can imagine allowing the user to only support a subset of such builtins. This is, in fact, what `cuecontext` is working towards, although there will need to be some more refactoring to make this possible.

**Comment by heistp** [comment]
Thanks, the apd package dependency does make more sense now.

I also realize that reducing or clarifying package dependencies is a wider and important challenge for the Go community as well. In my own code, I find the selection of package boundaries to be both consequential and not always trivial to get right.

**Comment by myitcv** [comment]
It's not totally clear there is anything to do here. I'll move this to be a discussion, and in case there are concrete issues/ideas that come from that discussion we can create appropriate issues.
